docs/man/mutool.1
.TH "MUTOOL" "1" "April 14, 2025"
.\" Please adjust this date whenever revising the manpage.
.\" no hyphenation
.nh
.\" adjust left
.ad l

.SH NAME
mutool \- all purpose tool for dealing with PDF files

.SH SYNOPSIS
mutool <sub-command> [options]

.SH VERSION
mutool -v
.PP
Shows the MuPDF version used to build mutool.

.SH DESCRIPTION
mutool is a tool based on MuPDF for dealing with document files in various manners.
There are several sub commands available, as described below.

.SH DRAW
mutool draw [options] file [pages]
.PP
The draw command will render a document to image files,
convert to another vector format, or extract the text content.
.PP
The supported input document formats are: pdf, xps, cbz, and epub.
.PP
The supported output image formats are: pbm, pgm, ppm, pam, png, pwg, pcl and ps.
The supported output vector formats are: svg, pdf, and debug trace (as xml).
The supported output text formats are: plain text, html, and structured text (as xml).
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-o output
The output format is inferred from the output filename.
Embed %d in the name to indicate the page number (for example: "page%d.png").
Printf modifiers are supported, for example "%03d".
If no output is specified, the output will go to stdout.
.TP
.B \-F format
Enforce a specific output format. Only necessary when outputting to stdout
since normally the output filename is used to infer the output format.
.TP
.B \-q
Be quiet, do not print progress messages.
.TP
.B \-R angle
Rotate clockwise by given number of degrees.
.TP
.B \-r resolution
Render the page at the specified resolution.
The default resolution is 72 dpi.
.TP
.B \-w width
Render the page at the specified width (or, if the -r flag is used,
render with a maximum width).
.TP
.B \-h height
Render the page at the specified height (or, if the -r flag is used,
render with a maximum height).
.TP
.B \-f
Fit exactly; ignore the aspect ratio when matching specified width/heights.
.TP
.B \-b  box
Use name page box (`MediaBox`, `CropBox`, `BleedBox`, `TrimBox`, or `ArtBox`), see :ref:`defined boxes<mutool_trim_defined_boxes>`.
.TP
.B \-B bandheight
Render in banded mode with each band no taller than the given height. This uses
less memory during rendering. Only compatible with pam, pgm, ppm, pnm and png
output formats. Banded rendering and md5 checksumming may not be used at the
same time.
.TP
.B \-T threads
Number of threads to use for rendering (banded mode only).
.TP
.B \-W width
Page width in points for EPUB layout.
.TP
.B \-H height
Page height in points for EPUB layout.
.TP
.B \-S size
Font size in points for EPUB layout.
.TP
.B \-U filename
User CSS stylesheet for EPUB layout.
.TP
.B \-X
Disable document styles for EPUB layout.
.TP
.B \-a
Disable usage of accelerator file.
.TP
.B \-c colorspace
Render in the specified colorspace.
Supported colorspaces are: \fBmono\fR, \fBgray\fR, \fBgrayalpha\fR, \fBrgb\fR, \fBrgbalpha\fR, \fBcmyk\fR, \fBcmykalpha\fR.
Some abbreviations are allowed: \fBm\fR, \fBg\fR, \fBga\fR, \fBrgba\fR, \fBcmyka\fR.
The default is chosen based on the output format.
.TP
.B \-e filename
Proof ICC profile filename to use for rendering.
.TP
.B -G gamma
Apply gamma correction.
Some typical values are 0.7 or 1.4 to thin or darken text rendering.
.TP
.B -I
Invert colors.
.TP
.B \-s [mft5]
Show various bits of information:
.B m
for glyph cache and total memory usage,
.B f
for page features such as whether the page is grayscale or color,
.B t
for per page rendering times as well statistics, and
.B 5
for md5 checksums of rendered images that can be used to check if rendering has
changed.
.TP
.B \-A bits
Specify how many bits of anti-aliasing to use. The default is 8.
.TP
.B \-A bits/bits
Specify how many bits of anti-aliasing to use. The default is 8. Bits for graphics is before the slash, bits for text after the slash.
.TP
.B \-l width
Minimum stroked line width (in pixels).
.TP
.B \-K
Do not draw text.
.TP
.B \-KK
Only draw text.
.TP
.B \-D
Disable use of display lists. May cause slowdowns, but should reduce
the amount of memory used.
.TP
.B \-i
Ignore errors.
.TP
.B \-m limit
Limit memory usage in bytes.
.TP
.B \-L
Low memory mode (avoid caching objects by clearing cache after each page).
.TP
.B \-P
Run interpretation and rendering at the same time.
.TP
.B \-N
Disable ICC workflow.
.TP
.B \-O overprint
Control spot/overprint rendering: \fB0\fR for no spot rendering, \fB1\fR for Overprint simulation (default), or \fB2f\R for full spot rendering.
.TP
.B \-t language
Specify language/script for OCR (default: eng)
.TP
.B \-d ocr-file-path
Specify path for OCR files (default: rely on \fBTESSDATA_PREFIX\fR environment variable.
.TP
.B \-k correction
Set the skew correction, either one of \fBauto\fR, \fB0\fR for increase size, \fB1\fR for maintain size, or \fB2\fR for decrease size.
.TP
.B \-k correction,angle
Set the skew correction as well as the angle.
.TP
.B \-y l
Print the layer configs to stderr.
.TP
.B \-y layer-number
Select layer config (by number from \fB-y l\fR).
.TP
.B \-y layer-number,item1,item2,...
Select layer config (by number from \fB-y l\fR) and toggle the listed items.
.TP
.B \-Y
Print the individual layers to stderr.
.TP
.B \-z layer-number
Hide individual layer.
.TP
.B \-Z layer-number
Show individual layer.

.TP
.B pages
Comma separated list of page numbers and ranges (for example: 1,5,10-15,20-N), where the character N denotes the last page.
If no pages are specified, then all pages will be rendered.

.SH CLEAN
mutool clean [options] input.pdf [output.pdf] [pages]
.PP
The clean command pretty prints and rewrites the syntax of a PDF file.
It can be used to repair broken files, expand compressed streams, filter
out a range of pages, etc.
.PP
If no output file is specified, it will write the cleaned PDF to "out.pdf"
in the current directory.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-g
Garbage collect objects that have no references from other objects.
Give the option twice to also renumber all objects and compact the cross reference table.
Give it three times to also merge and reuse duplicate objects.
Give it four times to also merge and reuse duplicate streams.
.TP
.B \-s
Rewrite content streams.
.TP
.B \-l
Linearize output. No longer supported! Used to ccreate a "Web Optimized" output file.
.TP
.B \-D
Save file without encryption.
.TP
.B \-E method
Save file with desired encryption method, one of: \fBrc4-40\fR \fBrc4-128\fR \fBaes-128\fR, or \fBaes-256\fR.
.TP
.B \-O password
Owner password used when encrypting output file. This is required to edit the output file.
.TP
.B \-U password
User password used when encrypting output file. This is required to view the output file.
.TP
.B \-P permissions
Permissions used when encrypting output file. These are the permissions that apply when the output file is viewed.
.TP
.B \-a
ASCII Hex encode binary streams. Use in conjunction with -d and -i or -f to
ensure that although the images and/or fonts are compressed, the resulting
file can still be viewed and edited with a text editor.
.TP
.B \-d
Decompress streams. This will make the output file larger, but provides
easy access for reading and editing the contents with a text editor.
.TP
.B \-z
Deflate uncompressed streams.
If combined with -d, any decompressed streams will be recompressed.
If combined with -a, the streams will also be hex encoded after compression.
.TP
.B \-e effort
Configure how much effort is spent compressing the output file. 0 is default amount, otherwise a percentage of the maximum effort, 1 for minimum and 100 for maximum effort.
.TP
.B \-f
Toggle decompression of font streams. Use in conjunction with -d to leave
fonts compressed.
.TP
.B \-i
Toggle decompression of image streams. Use in conjunction with -d to leave
images compressed.
.TP
.B \-c
Pretty-print graphics commands in content streams.
.TP
.B \-s
Sanitize graphics commands in content streams, this means, for example, removing invalid or redundant commands.
.TP
.B \-t
Use compact PDF object syntax. This saves space.
.TP
.B \-tt
Use indented PDF object syntax. This is more human-readable.
.TP
.B \-L
Print comments containing labels showing how each object can be reached from the Root.
.TP
.B \-A
Create appearance streams for annotations that are missing appearance streams.
.TP
.B \-AA
Recreate appearance streams for all annotations.
.TP
.B \-m
Preserve metadata.
.TP
.B \-S
Subset fonts if possible. (EXPERIMENTAL!)
.TP
.B \-Z
Use object streams cross reference streams for extra compressoin.
.TP
.B \-\-\fR(\fBcolor\fR,\fBgray\fR,\fBbitonal\fR)\fB-\fR(,\fBlossy-\fR,\fBlossless-\fR)\fBimage-subsample-method method
Set the subsampling method (\fBaverage\fR, or \fBbicubic\fR) for the
desired image types, for example color-lossy and bitonal-loessless.
.TP
.B \-\-\fR(\fBcolor\fR,\fBgray\fR,\fBbitonal\fR)\fB-\fR(,\fBlossy-\fR,\fBlossless-\fR)\fBimage-subsample-dpi dpi
Set the resolution at which to subsample.
.TP
.B \-\-\fR(\fBcolor\fR,\fBgray\fR,\fBbitonal\fR)\fB-\fR(,\fBlossy-\fR,\fBlossless-\fR)\fBimage-recompress-method quality
Set the recompression quality to either of \fBnever\fR, \fBsame\fR, \fBlossless\fR, \fBjpeg\fR, \fBj2k\fR, \fBfax\fR, or \fBjbig2\fR.
.TP
.B \-\-structure=\fR(\fBkeep\fR|\fBdrop\fR)
Keep or drop the structure tree.
.TP
.B pages
Comma separated list of page numbers and ranges (for example: 1,5,10-15,20-N), where the character N denotes the last page.
If no pages are specified, then all pages will be included.

.SH EXTRACT
mutool extract [options] file.pdf [object numbers]
.PP
The extract command can be used to extract images and font files from a PDF.
If no object numbers are given on the command line, all images and fonts
will be extracted.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-r
Convert images to RGB when extracting them.
.TP
.B \-a
Embed SMasks as alpha channel..
.TP
.B \-N
Do not use ICC color conversions.

.SH INFO
mutool info [options] file.pdf [pages]
.PP
The info command lists the resources used on each page in a PDF file.
The default is to list all resource types, but if one
or more flags are given, only the flagged types will be shown.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B -F
List fonts.
.TP
.B -I
List images.
.TP
.B -M
List page dimensions.
.TP
.B -S
List shadings.
.TP
.B -P
List patterns.
.TP
.B -X
List form and postscript XObjects.
.TP
.B -Z
List ZUGFeRD info.
.TP
.B pages
Comma separated list of page numbers and ranges (for example: 1,5,10-15,20-N), where the character N denotes the last page.
If no pages are specified, then all pages will be included.

.SH CREATE
mutool create [-o output.pdf] [options] page1.txt [page2.txt ...]
.PP
The create command creates a new PDF file with the contents created
from one or more input files containing graphics commands.
.TP
.B \-o output
If no output file is specified, it will write the created PDF to "out.pdf"
in the current directory.
.TP
.B page.txt
A page is created for each input file, with the contents of the file copied
into the content stream. Special comments in the input files are parsed to
define the page dimensions and font and image resources:
.PP
%%MediaBox 0 0 500 800
.br
%%Rotate 90
.br
%%Font Tm Times-Roman
.br
%%Font Fn0 path/to/font/file.ttf
.br
%%Image Im0 path/to/image.png
.TP
.B \-O options
Comma separated list of format specific output options, see \fIOUTPUT OPTIONS\fR below.

.SH PAGES
mutool pages [options] input.pdf [pages ...]
.PP
The pages command dumps information about the size and orientation
of pages within the document.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B pages
Comma separated list of page numbers and ranges (for example: 1,5,10-15,20-N), where the character N denotes the last page.
If no pages are specified, then all pages will be included.

.SH POSTER
mutool poster [options] input.pdf [output.pdf]
.PP
The poster command splits each page into tiles, and puts each tile on
a page of its own. It's useful for printing a large page onto smaller
pieces of paper that can then be glued together to create a large poster.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-m margin
Margin (overlap) between pages in points or percent.
.TP
.B \-x factor
Split the page into this many horizontal pieces.
.TP
.B \-y factor
Split the page into this many vertical pieces.
.TP
.B \-r
Splits hoirzontally from right to left (default splits from left to right).
.PP
The output will have x times y number of pages for each input page.

.SH SHOW
mutool show [options] file.pdf [object numbers ...]
.PP
The show command will print the specified objects and streams to stdout.
Streams are decoded and non-printable characters are represented
with a period by default.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-o file
Write output to file instead of stdout.
.TP
.B \-b
Print streams as binary data and omit the object header.
.TP
.B \-e
Print streams in their original encoded (or compressed) form.
.TP
.B \-g
Print each object on a single line, suitable for grep.
.TP
.B \-r
Force repair before showing any objects.
.PP
Specify objects by number, or use one of the following special names:
.TP
.B 'xref'
Print the cross reference table.
.TP
.B 'trailer'
Print the trailer dictionary.
.TP
.B 'encrypt'
Print the encryption dictionary.
.TP
.B 'pagetree'
List the object numbers for every page.
.TP
.B 'grep'
Print all the objects in the file in a compact one-line format suitable for piping to grep.
.TP
.B 'outline'
Print the outline (table of contents).

.SH RUN
mutool run script.js [arguments]
.PP
Executes a Javascript program which has access to most of the features of the
MuPDF library. The command supports ECMAScript 5 syntax in strict mode. All of
the MuPDF constructors and functions live in the global object, and the command
line arguments are accessible from the global argv object.
.PP
If invoke without any arguments, it will drop you into an interactive REPL
(read-eval-print-loop). On the interactive prompt, if you prefix a line with an
equal character it will automatically print the results of the line.
.PP
See the MuPDF documentation for details about the Javascript interfaces.

.SH CONVERT
mutool convert [options] file [pages]
.PP
The convert command is used to convert a file from one format to another.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-b  box
Use name page box (`MediaBox`, `CropBox`, `BleedBox`, `TrimBox`, or `ArtBox`), see :ref:`defined boxes<mutool_trim_defined_boxes>`.
.TP
.B \-A bits
Specify how many bits of anti-aliasing to use. The default is 8.
.TP
.B \-W width
Page width in points for EPUB layout.
.TP
.B \-H height
Page height in points for EPUB layout.
.TP
.B \-S size
Font size in points for EPUB layout.
.TP
.B \-U filename
User CSS stylesheet for EPUB layout.
.TP
.B \-o output
The output format is inferred from the output filename.
Embed %d in the name to indicate the page number (for example: "page%d.png").
Printf modifiers are supported, for example "%03d".
If no output is specified, the output will go to stdout.
.TP
.B \-F format
Enforce a specific output format. Only necessary when outputting to stdout
since normally the output filename is used to infer the output format.
.TP
.B \-O options
Comma separated list of format specific output options, see \fIOUTPUT OPTIONS\fR below.

.SH MERGE
mutool merge [options] file1 [pages] file2 [pages] ...
.PP
The merge command is used to pick out pages from two or more files and merge
them in order into a new PDF output file.
.TP
.B \-o output
The output filename.
.TP
.B \-O options
Comma separated list of format specific output options, see \fIOUTPUT OPTIONS\fR below.

.SH BAKE
mutool bake [options] input.pdf [output.pdf]
.PP
Normally annotations and form field widgets are separate content streams
that get drawn on top of the page contents. This command bakes annotations
and/or widgets into the page contents, making them inseparable.
After baking in annotations and widgets they are just visual representaitons
of what the original annotation/widget looked like, their properties can no
longer be edited since they no longer exist as proper annotation/widgets.
.TP
.B \-A
Do not bake in annotations.
.TP
.B \-F
Do not bake in form field widgets.
.TP
.B \-O options
A comma separated list of PDF options. See \fIOUTPUT OPTIONS\fR below.

.SH RECOLOR
mutool recolor [options] input.pdf
.PP
Change the color space used in a PDF to a desired one. This command can,
for example, change a color PDF to be grayscale.
.TP
.B \-c colorspace
Desired output colorspace, choose one of \fBgray\fR (default), \fBrgb\fR, or \fBcmyk\fR.
.TP
.B \-r
Remove any output intents.
.TP
.B \-o filename
The output filename.

.SH TRACE
mutool trace [options] file1 [pages] file2 [pages] ...
.PP
Prints a trace of device calls needed to render the desired pages from the given document.
.TP
.B \-p password
Use the specified password if the file is encrypted.
.TP
.B \-W width
Page width in points for EPUB layout.
.TP
.B \-H height
Page height in points for EPUB layout.
.TP
.B \-S size
Font size in points for EPUB layout.
.TP
.B \-U filename
User CSS stylesheet for EPUB layout.
.TP
.B \-X
Disable document styles for EPUB layout.
.TP
.B \-d
Enable use of display lists.
.B pages
Comma separated list of page numbers and ranges (for example: 1,5,10-15,20-N), where the character N denotes the last page.
If no pages are specified, then all pages will be rendered.

.SH TRIM

CHANGES
List of changes in MuPDF 1.27

	New and updated documentation! Build local HTML and Markdown versions from source with sphinx.

	New (experimental) search API that supports matching results across pages and regular expressions.

	Build system:
		Optimize thirdparty dependency submodules to reduce size of git clone and release tarballs.
		Allow conditional compilation for html, xps, svg, extract modules.
		Visual Studio project added an arm64 configuration.
		WASM bindings folded back into main git repository.
		Makefile added new "install-tools" target.

	General:
		More robust document type recognition.
		Many new classes and functions to both WASM and mutool run Javascript bindings.
		Changes to mutool run to bring it into sync with WASM library -- if you stick to ES5 syntax, scripts can run in both environments.
		Improve dehyphenation behavior in structured text extraction.
		Several optimizations when encountering JPEG2000 images in PDF files.
		Fixed many bugs with OCG handling (more improvements to come in a future release).

	HTML and EPUB:
		Hyphenation!
			Set HTML lang attribute and CSS style hyphens:auto to enable.
		CSS non-static positioning:
			position: absolute and relative
			inset/top/right/bottom/left
		HTML tables:
			colgroup and col
			rowspan and colspan
			width, height, valign, and border-collapse
		Legacy HTML attributes:
			width, height, valign, align, bgcolor, border, hidden
		CSS border-style: double, dotted, dashed, outset, inset, ridge, groove
		CSS tree-structural pseudo-selectors (:root, :empty, :first-child, :nth-child, etc)
		CSS :link pseudo-class
		Improved BiDi support with direction:rtl CSS property and dir=rtl HTML attribute
		Recognize all CSS named colors!
		Improved fallback font detection.
		EPUB 3 table of contents.

	PDF Forms & Annotations:
		Respect AcroForm/NeedAppearances flag -- widgets may look different if this flag is set!
		Changed Text (aka Note) annotation sizing and positioning to conform better with the PDF reference.
		Respect NoZoom and NoRotate flags.
		Fixed bug with some checkboxes not toggling as they should.

	Tools & Programs:
		mutool:
			Handle output to stdout more reliably in all tools.
		mutool run:
			allow import statements for mupdf and fs modules
			script compatibility tweaks with node
			process.argv
			process.exit
		mutool clean:
			vectorize text (turn text objects into paths) with -v flag
		mutool draw:
			behavior tweaked to be more consistent across output formats
		mutool poster:
			improve behavior when pages have been rotated
		mutool merge:
			preserve external (only) links when merging documents
		mutool extract:
			extract embedded files and attachments
		mutool grep:
			new tool to search for text in documents
		mupdf-gl:
			confirm before opening potentially dangerous external links.
		mupdf-x11:
			refuse to open potentially dangerous external links.

	API changes:
		Use fz_quad in fz_warp_pixmap, fz_autowarp_pixmap, and fz_detect_document.

	API news:
		- special redaction mode PDF_TEXT_REMOVE_INVISIBLE to strip only invisible text.
		- pdf_check_document (PDFDocument.check in JS and Java) function to scan the file for errors.
		- pdf_set_page_tree_cache to allow faster first-page loads at the cost of slower loading of all other pages.
		- fz_stext_page_details -- stext can now represent text from multiple pages.

List of changes in MuPDF 1.26

	Removed PDF linearization support:

		See https://artifex.com/blog/mupdf-removes-linearisation for
		more information.

	Object labels:

		Write a comment before numbered objects listing the possible
		ways it can be reached from the document root.

		Enable with mutool show -L, mutool clean -L, and PDF write
		option "labels".

		Library functions pdf_load_object_labels and pdf_label_object.

	Brotli compression:

		Experimental feature from latest PDF specification proposal.

	Optional barcode support:

		Build with "make barcode=yes".
		Uses zxing-cpp (zebra crossing) C++ as a third party dependency.
		Adds mutool barcode sub-command.

		To detect barcodes in a PDF:
			mutool barcode -d barcode.pdf

		To create barcode images as PNG file:
			mutool barcode -c -o barcode.png -F qrcode "Hello, world!"

		Library functions fz_new_barcode_image and
		fz_decode_barcode_from_pixmap.

	New Structured Text extraction passes:

		Paragraph Breaking: Detect paragraphs from first-line
		indentation and other heuristics.

		Table Hunt: Scan text and try to assemble it into tables with
		rows and columns.

	New Structured Text extraction options:
		- allow reading raw character or glyph index when unicode is not available
		- collect styles to detect fake bold, underlines, strike-through, etc.
		- clip-rect to only extract text inside the specified area
		- accurate-ascenders to measure ascender data from font outline
		- accurate-side-bearings to measure glyph widths from font outline

	CSV output format (using "Table Hunt" pass on input document).

	Core library additions:

		JSON parser, object model, and printer.

		Activity logger API.

		Support "overlong null" characters in utf-8 (using "\xC0\x80"
		to represent \x00 in C strings, without being misconstrued as a
		string terminator).

	Build system changes:

		Cleaned up Makefile cruft and refactored how shared libraries
		are built.

		Removed bit-rotted sections of the Makefile that have not been
		maintained and no longer work (primarily iOS and MinGW).

		Don't build example tools muraster and mupdf-x11-curl by
		default. These examples can be built with "make extra-apps".

	Miscellaneous improvements and new behavior:

		Include more data in structured text XML dumps.

		Ignore junk at the start of a PDF file (such as PJL commands)
		without needing to perform a full repair pass.

		Support non-latin/greek/cyrillic/CJK languages in annotations
		and widgets (requires HTML layout engine).

		Use system font callback to look for missing fonts in HTML
		layout.


scripts/mutool.py
#!/usr/bin/env python

'''
Intended to behave exactly like mutool, but uses the mupdf python => C++ =>
mupdf.so wrappers.

The code is intended to be similar to the mutool C code, to simplify
comparison.
'''

import getopt
import os
import sys
import textwrap

if os.environ.get('MUPDF_PYTHON') in ('swig', None):
    # PYTHONPATH should have been set up to point to a build/shared-*/
    # directory containing mupdf.so generated by scripts/mupdfwrap.py and SWIG.
    import mupdf
elif os.environ.get('MUPDF_PYTHON') == 'cppyy':
    sys.path.insert(0, os.path.abspath(f'{__file__}/../../platform/python'))
    import mupdf_cppyy
    del sys.path[0]
    mupdf = mupdf_cppyy.cppyy.gbl.mupdf
else:
    raise Exception(f'Unrecognised $MUPDF_PYTHON: {os.environ.get("MUPDF_PYTHON")}')

def usage():
    print( textwrap.dedent('''
            usage: mutool.py <command> [options]
            \tclean\t-- rewrite pdf file
            \tconvert\t-- convert document
            \ttrace\t-- trace device calls
            \tdraw\t-- convert document
            '''))


# Things for clean
#
def clean_usage():
    print(textwrap.dedent(
            f'''
            usage: mutool clean [options] input.pdf [output.pdf] [pages]
            \t-p -\tpassword
            \t-g\tgarbage collect unused objects
            \t-gg\tin addition to -g compact xref table
            \t-ggg\tin addition to -gg merge duplicate objects
            \t-gggg\tin addition to -ggg check streams for duplication
            \t-l\tlinearize PDF
            \t-D\tsave file without encryption
            \t-E -\tsave file with new encryption (rc4-40, rc4-128, aes-128, or aes-256)
            \t-O -\towner password (only if encrypting)
            \t-U -\tuser password (only if encrypting)
            \t-P -\tpermission flags (only if encrypting)
            \t-a\tascii hex encode binary streams
            \t-d\tdecompress streams
            \t-z\tdeflate uncompressed streams
            \t-f\tcompress font streams
            \t-i\tcompress image streams
            \t-c\tclean content streams
            \t-s\tsanitize content streams
            \t-A\tcreate appearance streams for annotations
            \t-AA\trecreate appearance streams for annotations
            \tpages\tcomma separated list of page numbers and ranges
            '''
            ))
    sys.exit(1)

def clean(argv):
    outfile = 'out.pdf'
    password = ''
    opts = mupdf.PdfCleanOptions()
    opts.write.do_garbage += 1
    errors = 0
    items, argv = getopt.getopt( argv, 'adfgilp:sczDAE:O:U:P:')
    for option, value in items:
         if 0:   pass   # lgtm [py/unreachable-statement]
         elif option == '-p': password = value
         elif option == '-d': opts.write.do_decompress += 1
         elif option == '-z': opts.write.do_compress += 1
         elif option == '-f': opts.write.do_compress_fonts += 1
         elif option == '-i': opts.write.do_compress_images += 1
         elif option == '-a': opts.write.do_ascii += 1
         elif option == '-g': opts.write.do_garbage += 1
         elif option == '-l': opts.write.do_linear += 1
         elif option == '-c': opts.write.do_clean += 1
         elif option == '-s': opts.write.do_sanitize += 1
         elif option == '-A': opts.write.do_appearance += 1
         elif option == '-D': opts.write.do_encrypt = PDF_ENCRYPT_NONE
         elif option == '-E': opts.write.do_encrypt = encrypt_method_from_string(value)
         elif option == '-P': opts.write.permissions = int(value)
         elif option == '-O': opts.write.opwd_utf8 = value[:128]
         elif option == '-U': opts.write.upwd_utf8 = value[:128]
         else:
            clean_usage()

    if (opts.write.do_ascii or opts.write.do_decompress) and not opts.write.do_compress:
        opts.write.do_pretty = 1

    if not argv:
        clean_usage()

    infile = argv.pop(0)

    if argv and '.pdf' in argv[0].lower():
        outfile = argv.pop(0)

    try:
        mupdf.pdf_clean_file(infile, outfile, password, opts, argv)
    except Exception as e:
        print( f'mupdf.pdf_clean_file() failed: {e}')
        errors += 1
        if 0:
            # Enable for debugging.
            import traceback
            traceback.print_exc()
    return errors != 0;



# Things for draw.
#

import mutool_draw

draw = mutool_draw.draw



# Things for convert.
#


def convert_usage():
    print( textwrap.dedent(
            f'''
            mutool convert version {mupdf.FZ_VERSION}
            Usage: mutool convert [options] file [pages]
            \t-p -\tpassword

            \t-A -\tnumber of bits of antialiasing (0 to 8)
            \t-W -\tpage width for EPUB layout
            \t-H -\tpage height for EPUB layout
            \t-S -\tfont size for EPUB layout
            \t-U -\tfile name of user stylesheet for EPUB layout
            \t-X\tdisable document styles for EPUB layout

            \t-o -\toutput file name (%d for page number)
            \t-F -\toutput format (default inferred from output file name)
            \t\t\traster: cbz, png, pnm, pgm, ppm, pam, pbm, pkm.
            \t\t\tprint-raster: pcl, pclm, ps, pwg.
            \t\t\tvector: pdf, svg.
            \t\t\ttext: html, xhtml, text, stext.
            \t-O -\tcomma separated list of options for output format

            \tpages\tcomma separated list of page ranges (N=last page)
            '''
        ))
    print( mupdf.fz_draw_options_usage)
    print( mupdf.fz_pcl_write_options_usage)
    print( mupdf.fz_pclm_write_options_usage)
    print( mupdf.fz_pwg_write_options_usage)
    print( mupdf.fz_stext_options_usage)
    print( mupdf.fz_pdf_write_options_usage)
    print( mupdf.fz_svg_write_options_usage)
    sys.exit(1)


def convert_runpage( doc, number, out):
    page = mupdf.FzPage( doc, number - 1)
    mediabox = page.fz_bound_page()
    dev = out.fz_begin_page(mediabox)
    page.fz_run_page( dev, mupdf.FzMatrix(mupdf.fz_identity), mupdf.FzCookie())
    out.fz_end_page()

def convert_runrange( doc, count, range_, out):
    start = None
    end = None
    while 1:
        range_, start, end = mupdf.fz_parse_page_range( range_, count)
        if range_ is None:
            break
        step = +1 if end > start else -1
        for i in range( start, end, step):
            convert_runpage( doc, i, out)

def convert( argv):
    # input options
    password = ''
    alphabits = 8
    layout_w = mupdf.FZ_DEFAULT_LAYOUT_W
    layout_h = mupdf.FZ_DEFAULT_LAYOUT_H
    layout_em = mupdf.FZ_DEFAULT_LAYOUT_EM
    layout_css = None
    layout_use_doc_css = 1

    # output options
    output = None
    format_ = None
    options = ''

    items, argv = getopt.getopt( argv, 'p:A:W:H:S:U:Xo:F:O:')
    for option, value in items:
        if 0: pass  # lgtm [py/unreachable-statement]
        elif option == '-p':    password = value
        elif option == '-A':    alphabits = int(value)
        elif option == '-W':    layout_w = float( value)
        elif option == '-H':    layout_h = float( value)
        elif option == '-S':    layout_em = float( value)
        elif option == '-U':    layout_css = value
        elif option == '-X':    layout_use_doc_css = 0
        elif option == '-o':    output = value
        elif option == '-F':    format_ = value
        elif option == '-O':    options = value
        else:   assert 0

    if not argv or (not format_ and not output):
        convert_usage()

    mupdf.fz_set_aa_level( alphabits)
    if layout_css:
        buf = mupdf.FzBuffer( layout_css)
        mupdf.fz_set_user_css( buf.string_from_buffer())

    mupdf.fz_set_use_document_css(layout_use_doc_css)

    if format_:
        out = mupdf.FzDocumentWriter( output, format_, options)
    else:
        out = mupdf.FzDocumentWriter( output, options, mupdf.FzDocumentWriter.OutputType_PDF)

    i = 0
    while 1:
        if i >= len( argv):
            break
        arg = argv[i]
        doc = mupdf.FzDocument( arg)
        if doc.fz_needs_password():
            if not doc.fz_authenticate_password( password):
                raise Exception( f'cannot authenticate password: {arg}')
        doc.fz_layout_document( layout_w, layout_h, layout_em)
        count = doc.fz_count_pages()

        range_ = '1-N'
        if i + 1 < len(argv) and mupdf.fz_is_page_range(ctx, argv[i+1]):
            i += 1
            range_ = argv[i]
        convert_runrange( doc, count, range_, out)
        i += 1

    out.fz_close_document_writer()



# Things for trace.
#

def trace_usage():
    print( textwrap.dedent('''
            Usage: mutool trace [options] file [pages]
            \t-p -\tpassword

            \t-W -\tpage width for EPUB layout
            \t-H -\tpage height for EPUB layout
            \t-S -\tfont size for EPUB layout
            \t-U -\tfile name of user stylesheet for EPUB layout
            \t-X\tdisable document styles for EPUB layout

            \t-d\tuse display list

            \tpages\tcomma separated list of page numbers and ranges
            '''))
    sys.exit( 1)

def trace_runpage( use_display_list, doc, number):
    page = mupdf.FzPage( doc, number-1)
    mediabox = page.fz_bound_page()
    print( f'<page number="{number}" mediabox="{mediabox.x0} {mediabox.y0} {mediabox.x1} {mediabox.y1}">')
    output = mupdf.FzOutput( mupdf.FzOutput.Fixed_STDOUT)
    dev = mupdf.FzDevice( output)
    if use_display_list:
        list_ = mupdf.FzDisplayList( page)
        list_.fz_run_display_list( dev, mupdf.FzMatrix(mupdf.fz_identity), mupdf.FzRect(mupdf.fz_infinite_rect), mupdf.FzCookie())
    else:
        page.fz_run_page( dev, mupdf.FzMatrix(mupdf.fz_identity), mupdf.FzCookie())
    output.fz_close_output()
    print( '</page>')

def trace_runrange( use_display_list, doc, count, range_):
    start = None
    end = None
    while 1:
        range_, start, end = mupdf.fz_parse_page_range( range_, count)
        print(f'range_={range_!r} start={start} end={end}')
        if range_ is None:
            break
        step = +1 if end > start else -1
        for i in range( start, end, step):
            trace_runpage( use_display_list, doc, i)

def trace( argv):

    password = ''
    layout_w = mupdf.FZ_DEFAULT_LAYOUT_W
    layout_h = mupdf.FZ_DEFAULT_LAYOUT_H
    layout_em = mupdf.FZ_DEFAULT_LAYOUT_EM
    layout_css = None
    layout_use_doc_css = 1

    use_display_list = 0

    argv_i = 0
    while 1:
        arg = argv[ argv_i]
        if arg == '-p':
            password = next( opt)
        elif arg == '-W':
            argv_i += 1
            layout_w = float( argv[argv_i])
        elif arg == '-H':
            argv_i += 1
            layout_h = float( argv[argv_i])
        elif arg == '-S':
            argv_i += 1
            layout_em = float( argv[argv_i])
        elif arg == '-U':
            argv_i += 1
            layout_css = argv[argv_i]
        elif arg == '-X':
            layout_use_doc_css = 0
        elif arg == '-d':
            use_display_list = 1
        else:
            break
        argv_i += 1

    if argv_i == len( argv):
        trace_usage()

    if layout_css:
        buffer_ = mupdf.FzBuffer( layout_css)
        mupdf.fz_set_user_css( buffer_.string_from_buffer())

    mupdf.fz_set_use_document_css( layout_use_doc_css)

    for argv_i in range( argv_i, len( argv)):
        arg = argv[ argv_i]
        doc = mupdf.FzDocument( arg)
        if doc.fz_needs_password():
            doc.fz_authenticate_password( password)
        doc.fz_layout_document( layout_w, layout_h, layout_em)
        print( f'<document filename="{arg}">')
        count = doc.fz_count_pages()
        if argv_i + 1 < len( argv) and mupdf.fz_is_page_range( argv[ argv_i+1]):
            argv_i += 1
            trace_runrange( use_display_list, doc, count, argv[ argv_i])
        else:
            trace_runrange( use_display_list, doc, count, '1-N')
        print( '</document>')



def main( argv):
    arg1 = argv[1]
    fn = getattr( sys.modules[__name__], arg1, None)
    if not fn:
        print( f'cannot find {arg1}')
        usage()
        sys.exit(1)

    return fn( argv[2:])


if __name__ == '__main__':
    try:
        e = main( sys.argv)
        sys.exit(e)
    except Exception as e:
        if 0:   # Enable when debugging.
            sys.stdout.flush()
            sys.stderr.flush()
            print(f'Exception: {e}')
            sys.stdout.flush()
        raise
